admin-dev/themes/new-theme/js/pages/product/components/categories/category-tree-selector.ts

```diff
@@ -0,0 +1,471 @@
+        });
+      }
+    }, this);
+    // Tree is initialized we can show it and hide loader
+    if (this.modalContainer) {
+      const fieldset = this.modalContainer.querySelector(ProductCategoryMap.fieldset) as HTMLElement;
+      const loader = this.modalContainer.querySelector(ProductCategoryMap.loader) as HTMLElement;
+
+      fieldset.classList.remove('d-none');
+      loader.classList.add('d-none');
+    }
+  }
+
+  /**
+   * Used to recursively create items of the category tree
+   */
+  private generateCategoryTree(treeCategory: TreeCategory): HTMLElement {
+    const categoryNode = this.generateTreeElement(treeCategory) as HTMLElement;
+    const childrenList = categoryNode.querySelector(ProductCategoryMap.childrenList) as HTMLElement;
+    childrenList.classList.add('d-none');
+
+    const hasChildren = treeCategory.children && treeCategory.children.length > 0;
+    categoryNode.classList.toggle('more', hasChildren);
+    if (hasChildren) {
+      const inputsContainer = categoryNode.querySelector(ProductCategoryMap.treeElementInputs) as HTMLElement;
+
+      inputsContainer.addEventListener('click', (event) => {
+        // We don't want to mess with the inputs behaviour (no toggle when checkbox or radio is clicked)
+        // So we only toggle when the div itself is clicked.
+        if (event.target !== event.currentTarget) {
+          return;
+        }
+
+        const isExpanded = !childrenList.classList.contains('d-none');
+        categoryNode.classList.toggle('less', !isExpanded);
+        categoryNode.classList.toggle('more', isExpanded);
+        childrenList.classList.toggle('d-none', isExpanded);
+      });
+
+      // Recursively build the children trees
+      treeCategory.children.forEach((childCategory) => {
+        const childTree = this.generateCategoryTree(childCategory);
+
+        childrenList.append(childTree);
+      });
+    }
+
+    return categoryNode;
+  }
+
+  /**
+   * If the category is among the initial ones (inserted by the form on load) the existing element is used,
+   * if not then it is generated based on the prototype template. In both case the element is injected with the
+   * category name.
+   */
+  private generateTreeElement(category: TreeCategory): HTMLElement {
+    if (!this.prototypeTemplate || !this.prototypeName) {
+      throw 'Invalid category tree prototype template or name';
+    }
+
+    const template = this.prototypeTemplate.replace(new RegExp(this.prototypeName, 'g'), String(category.id));
+    // Trim is important here or the first child could be some text (whitespace, or \n)
+    const frag = document.createRange().createContextualFragment(template.trim());
+    const categoryNode = frag.firstChild as HTMLElement;
+
+    // Add category name text
+    const checkboxInput = categoryNode.querySelector(ProductCategoryMap.checkboxInput) as HTMLInputElement;
+    const nameElement = document.createTextNode(category.name);
+    const element = category.active ?
```

Comment:
I'm not sure I understand this part.. If the category is not active it is displayed in the tree but in italic?